### ISSUE TITLE
Prefix in @ConfigProperties

### ISSUE THREAD

**Issue by radcortez** [eclipse/microprofile-config#588]
**Describe the bug**

I think we need to revise the way that we specced the behaviour of prefix in @ConfigProperties. Right now the prefix does not behave like other implementations of configurations (like Spring, Micronaut or Quarkus).

We force the user to add the dot in the end of the configuration prefix. This is counter intuitive, compared to the other platforms and even other APIs with similar behaviours. 

We should fix this and specify that the ending dot is not required in the prefix to perform the lookup.

**Comment by kenfinnigan** [comment]
+1

Anything hierarchical I've seen, configuration or otherwise, doesn't require the dot to be present.

**Comment by jclingan** [comment]
I am the instigator behind this issue when I discovered it reading the updated Config 2.0 spec. This caught me off guard since not auto-including the "dot" in the property is very counter-intuitive. The ship has sailed - developers assume the "." will be added based on how other Java runtimes work. IMHO, the spec needs to be updated to align with industry expectations.

**Comment by Emily-Jiang** [comment]
The reason to spell out the prefix exactly is to cater for other separator might be used. Look at the following scenario;
```
customer_age=21
customer_name=David
```

What prefix do we ask the end user to specify? Clearly we should not automatically add dot.

**Comment by radcortez** [comment]
If that is the case, implementations could define a configurable separator char. 

I understand that there are cases like this, but probably a small minority. Is it worth to confuse everybody to remember they need to add the separator?

**Comment by Emily-Jiang** [comment]
I thought this further and think the following solution should fix most cases.

```
public @interface ConfigProperties {
    @Nonbinding
    String prefix() default "";
//introduce the following method
   @Nonbinding
    String exactPrefix() default "";

```
In this case, the prefix can automatically add dot where the fullPrefix is to cater for all kinds of other prefix, such as
```
customer$name=Bob
customer$age=21
```
or 
```
customerName=Ben
customerAge=21
```
In the above scenario, you can do`@ConfigProperties(exactPrefix="customer")` or `@ConfigProperties(exactPrefix="customer$")`.

**Comment by kenfinnigan** [comment]
Are there use case in Java already where this mixture of separators is used?

Or is it purely to be flexible?

**Comment by Emily-Jiang** [comment]
I think it is flexible and there are cases with a separator instead of dot.

**Comment by dmlloyd** [comment]
Using a dot separator is two things:

* The standard separator for the API side
* The separator used in certain configuration sources (properties, system properties)

Other configuration sources don't use dot for the separator on the back end.  But everything uses dot as the separator on the front end.  Roberto is right that requiring a trailing dot is not useful to users.  Supporting "custom" separators by way of the prefix is at best a weird transgression of encapsulation, and at worst something that can't be supported by non-properties-based configuration sources.  Flexibility is useful in some cases, but ignoring the lines between abstractions is not flexibility.  In fact it ends up having the opposite effect.

**Comment by Emily-Jiang** [comment]
How can you be so sure to assert everything uses dot as the separator on the front end? I think with what I suggested addresses most of use cases. I don't get supporting "custom" separator being weird transgression of encapsulation.

**Comment by dmlloyd** [comment]
> How can you be so sure to assert everything uses dot as the separator on the front end?

It is in the specification, and in every usage example!

**Comment by radcortez** [comment]
@Emily-Jiang considering that other runtimes like Spring and Micronaut add the dot, shouldn't we follow the same behaviour? 

For me the issue here is not who is right or wrong, but about convenience. A great number of developers is already used to this behaviour, so why confuse them? And, nothing stops the implementations to provide a different behaviour based on some configuration.

**Comment by Emily-Jiang** [comment]
I don't have objections to add dot but also want to support other separators not limiting ourselves. 
What is wrong with what I suggested [here](https://github.com/eclipse/microprofile-config/issues/588#issuecomment-668585059). It addresses the request from @jclingan ! 

> > How can you be so sure to assert everything uses dot as the separator on the front end?
> 
> It is in the specification, and in every usage example!

This spec does not represent every usage example. This spec does not define the separator for config names either. It can be anything.

**Comment by Teg79** [comment]
I think also in this case the principle "Convention over Configuration" can be applied.

This is my idea
Convention: use the dot as properties separator, don't repeat it in every prefix.
Configuration: add another annotation property to specify you custom separator.

**Comment by Emily-Jiang** [comment]
I am happy with what @Teg79 suggested. @radcortez do you want to update your PR accordingly? Just more explaination on why we need this. At the moment, we can do this:
```
@Inject @ConfigProperty(name="customer_name") String name;
@Inject @ConfigProperty(name="customer_age") int age;

``` 
If we only support the dot as the separator for `@ConfigProperties`, we can't use `ConfigProperties` to cover the above scenario as we will search for `customer.name` and `customer.age`.

```

@ConfigProperties(prefix="customer")
@Dependent
public class MyCustomer {

public String name;
public int age;
}

```

**Comment by kenfinnigan** [comment]
Is the concern purely that if existing applications currently use `customer_name`, from your example, and they want to switch to a config object that they'd be forced to change the separator in their properties files?

**Comment by radcortez** [comment]
I'm not fully convinced with the separator support idea.

Since we cannot tell which separator to use until the user ask for a mapping, implementations will have to split the keys every time a mapping is performed. Sure, this can be cached, but it would still require for everything to happen at runtime. You are not able to validate the keys either.

Also, the suggested approach opens the door to have mix-ins. Nothing prevents you from having a mapping with dot separator, another one with underscore, and many others with different separators. This just seems a recipe for the user to shoot himself in the foot with so many options.

If we really want to support something like this, it has to be globally. It has to be set as the Config separator to use on a new instance creation.

I still think is too much work for little gain and definitely something that can be covered at the implementation side.

**Comment by jclingan** [comment]
I think @kenfinnigan brings up a good point. @Emily-Jiang Can you point to real-world (non-hypothetical) examples where this is an issue? It feels to me like we are optimizing for an edge use case.

A compromise is to use the "." separator as the default, but my preference is to assume just the "." separator and move forward.  In MicroProfile, we iterate quickly and can always come back and update the spec when we get real-world feedback otherwise.

**Comment by radcortez** [comment]
BTW, we already support this in SmallRye Config via a Relocator, which allows you to provide a function to relocate your config names.

**Comment by Emily-Jiang** [comment]
> Is the concern purely that if existing applications currently use `customer_name`, from your example, and they want to switch to a config object that they'd be forced to change the separator in their properties files?

If we fix the separator to ., yes, for the customers use `customer_name` cannot use the `ConfigProperties` to aggregate the properties. 
By the way, [DeltaSpike](https://deltaspike.apache.org/documentation/configuration.html) does what MP Config currently has:

```
If all your keys use the same prefix you can configure it on @Configuration:

@Configuration(prefix = "client.")
public interface MyConfig {
    @ConfigProperty(name = "url")
    String url();

    @ConfigProperty(name = "timeout", defaultValue = "30000")
    long timeout();
}

```
In MP Config, we don't define a rule on what is allowed in config name. In `ConfigProperty` we support all config names. Why does `@ConfigProperties` only force people to separate the parts in the name using dot. it is inconsistent. Besides, it is so easy to provide the parity.

**Comment by Emily-Jiang** [comment]
> I think @kenfinnigan brings up a good point. @Emily-Jiang Can you point to real-world (non-hypothetical) examples where this is an issue? It feels to me like we are optimizing for an edge use case.

A compromise is to use the "." separator as the default, but my preference is to assume just the "." separator and move forward. In MicroProfile, we iterate quickly and can always come back and update the spec when we get real-world feedback otherwise.

As explained earlier, this creates inconsistence between `@ConfigProperty` and `@ConfigProperties', which should just an aggregation on `@ConfigProperty`.

**Comment by kenfinnigan** [comment]
A couple of points from my perspective:

- I don't like the DeltaSpike approach as it is prone to errors with people forgetting to add a separator to the root config name. I've never seen it done this way anywhere else
- I don't see it as being unreasonable to say that if a developer is converting from a large number of common prefixed configuration names `customer_`, and thus creating a specific config class to capture the root, that they'd also need to change the separator.

My preference would be for `.` as default. I'm uncertain about letting it be customized

**Comment by radcortez** [comment]
Like @jclingan says, we could use use dot as a default for now and came back later and add customisation if required in a future version.

**Comment by Teg79** [comment]
> 
> Since we cannot tell which separator to use until the user ask for a mapping, implementations will have to split the keys every time a mapping is performed. Sure, this can be cached, but it would still require for everything to happen at runtime. You are not able to validate the keys either.
> 

@radcortez This is not clear for me, if we use an annotation method to specify the separator we can know the full prefix with separator at build time, I cannot see the difference between a single annotation method and concatenating 2 of them.

On the other side I must say that in migrating an application to MicroProfile there is no problem to change any separator on a legacy properties file, because we have to rethink the full structure and also the configuration names and where to place them. 

If the choice is between an implicit dot separator or to express the separator in every prefix then I think the first is better and safer. In this way we need only to adapt the configs/env variables that is in the Ops area leaving the Devs with a more intuitive approach.

**Comment by radcortez** [comment]
> > Since we cannot tell which separator to use until the user ask for a mapping, implementations will have to split the keys every time a mapping is performed. Sure, this can be cached, but it would still require for everything to happen at runtime. You are not able to validate the keys either.
> 
> @radcortez This is not clear for me, if we use an annotation method to specify the separator we can know the full prefix with separator at build time, I cannot see the difference between a single annotation method and concatenating 2 of them.
> 

MP Config may be used as a standalone API, without a CDI container or in frameworks / apps that don't do build time annotation processing, so we need to think in multiple scenarios and how the spec may behave with different runtimes.

> If the choice is between an implicit dot separator or to express the separator in every prefix then I think the first is better and safer. In this way we need only to adapt the configs/env variables that is in the Ops area leaving the Devs with a more intuitive approach.

My preference is also to have an implicit separator. Dot is the obvious candidate since its the one we use in the spec examples and TCK, plus it is widely adopted across the industry. 

I'm not against adding a way to define other separators, but I still think that if we provide a way for it on a `ConfigProperties` basis, it may open the door to some confusing scenarios. Consider:

``` 
#ConfigSource A
my.prop=1234
```

```
#ConfigSource B
my_prop=5678
```

Are `my.prop` and `my_prop` the same property? In theory, no. In practice, I see frustrated people trying to figure out why they are not getting the right value from their configuration :)

In conclusion, if we want to support this, I believe this needs to be global to the `Config`. Or, we could do this properly and instead of thinking into separators, we think just about names and transformations. It may be more interesting to support a feature that supports the "Relocation" of a Configuration set. Something like: `configuration lookup -> relocation function -> configuration retrieval`. This would allow to use the same configuration name structure when performing lookups across the code, and the relocation function could replace the separator for something else, and even only in certain cases.

**Comment by Emily-Jiang** [comment]
> I'm not against adding a way to define other separators, but I still think that if we provide a way for it on a ConfigProperties basis, it may open the door to some confusing scenarios. Consider:

#ConfigSource A
my.prop=1234

#ConfigSource B
my_prop=5678

Are my.prop and my_prop the same property? In theory, no. In practice, I see frustrated people trying to figure out why they are not getting the right value from their configuration :)

In conclusion, if we want to support this, I believe this needs to be global to the Config. Or, we could do this properly and instead of thinking into separators, we think just about names and transformations. It may be more interesting to support a feature that supports the "Relocation" of a Configuration set. Something like: configuration lookup -> relocation function -> configuration retrieval. This would allow to use the same configuration name structure when performing lookups across the code, and the relocation function could replace the separator for something else, and even only in certain cases.

In my understanding `my.prop` is different from `my_prop`. We should not lump this. The name is different. Even if they end up inside enivornment variable, if you look up `mp.prop`, you will get `1234` if both are present.

**Comment by radcortez** [comment]
Yes the names are different and I stated just that. But they are also similar, in the sense that they may be a mistake from the developers and we should help to avoid these mistakes. 

The only thing I'm advocating is that we should not support multiple separator for different configurations in the same `Config` instance. I see the separator as the profile, you want to use a different one, sure, set it up at start, but not on demand. We don't have on demand separators for File systems do we ? :)

**Comment by Emily-Jiang** [comment]
> The only thing I'm advocating is that we should not support multiple separator for different configurations in the same Config instance. I see the separator as the profile, you want to use a different one, sure, set it up at start, but not on demand. We don't have on demand separators for File systems do we ? :)
I think it is different from File system. Anyway, I think the MP Config implementation can add this kind of restriction if they wish. I don't think MP Config should do this. I am not aware of this kind of rules from Spring Config or DeltaSpike Config. I will send an email to the google group mailinglist so that more people can share their voices. We can then go with majority.

**Comment by Emily-Jiang** [comment]
Option 1:
Use dot as default and specify the separator otherwise.

In most cases, dot is probably the most popular separator. We plan to default the separator the dot. If a different separator is used, you can specify the separator using separator =“_”.

@ConfigProperties(prefix=“customer”)  

or @ConfigProperties(prefix=“customer”, separator =“_”)

Pro: very flexible 
Cons: two parameters to specify if dot is not the separator.

**Comment by Emily-Jiang** [comment]
Option 2:
Only support dot as the separator as it might be the most popular separator. If the prefix and the remaining names use a different separator, the prefix can’t be used.
Props: simple 
Cons: 
1. limited usage and it doesn’t cover the scenario where the property names don’t use dot as a separator. 
2. It is not consistent with @ConfigProperty where there is no restriction on config names.

**Comment by dmlloyd** [comment]
The only question you need to ask is, "what happens when I have a truly hierarchical configuration source?" and you'll see that having "flexible" separators is false flexibility.  Separators are only relevant to configuration sources that are not inherently hierarchical, like properties files.  Once you realize this, it becomes obvious that you need to have a standardized separator for different situations:

* When specifying the name (or, to use a more accurate term, "path") of a configuration property to find by way of one of the APIs
* When mapping a hierarchical name to a flat namespace like properties files

Here's the thing though.  In the former case, it is illogical to allow more than one separator (because it doesn't necessarily carry through to the configuration source when the configuration source is already hierarchical).  In the latter case, the best separator depends completely on the implementation - for example, properties files use `.` by *convention*, but filesystems use `/` or `\` by *requirement*.  Other formats may have other native separator characters.  By blurring the purpose of the separator between the front and back ends are you are doing (by only considering the properties file case), you're just confusing yourself.  You *think* you're adding flexibility to the user, but instead you're actually *removing* flexibility from backends and implementers, which will cost the user much more than being able to specify a separator of `!` or `*` or whatever, which is of limited utility anyway.

In summary, *require* the usage of a `.` separator on the front end, and specify that the configuration sources should map this separator to the most correct separation concept that the particular configuration source supports.  If the user wants a customized separator in their properties file, let them implement an alternative `ConfigSource` with that behavior.  Establish clear lines of encapsulation, and *stick with them*!
